include/libp2p/transport/tcp/tcp_util.hpp

```diff
@@ -6,134 +6,200 @@
 
 #pragma once
 
-#include <sstream>
-#include <system_error>  // for std::errc
-
-#include <boost/asio.hpp>
-#include <boost/lexical_cast.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ip/udp.hpp>
+#include <charconv>
 #include <libp2p/multi/multiaddress.hpp>
 
 namespace libp2p::transport::detail {
-  template <typename T>
-  inline outcome::result<multi::Multiaddress> makeAddress(
-      T &&endpoint, const ProtoAddrVec *layers = nullptr) {
-    try {
-      auto address = endpoint.address();
-      auto port = endpoint.port();
-
-      // TODO(warchant): PRE-181 refactor to use builder instead
-      std::ostringstream s;
-      if (address.is_v4()) {
-        s << "/ip4/" << address.to_v4().to_string();
-      } else {
-        s << "/ip6/" << address.to_v6().to_string();
+  using P = multi::Protocol::Code;
+
+  struct Dns {
+    std::optional<bool> v4;
+    std::string name;
+  };
+  using IpOrDns = std::variant<boost::asio::ip::address, Dns>;
+
+  static outcome::result<IpOrDns> readIpOrDns(ProtoAddrVec::iterator &it,
+                                              ProtoAddrVec::iterator end) {
+    if (it == end) {
+      return std::errc::protocol_not_supported;
+    }
+    auto &[p, v] = *it++;
+    switch (p.code) {
+      case P::IP4:
+      case P::IP6: {
+        boost::asio::ip::address ip;
+        boost::system::error_code ec;
+        switch (p.code) {
+          case P::IP4:
+            ip = boost::asio::ip::make_address_v4(v, ec);
+            break;
+          case P::IP6:
+            ip = boost::asio::ip::make_address_v6(v, ec);
+            break;
+        }
+        if (ec) {
+          return ec;
+        }
+        return ip;
       }
-
-      s << "/tcp/" << port;
-
-      if (layers != nullptr and not layers->empty()) {
-        auto &protocol = layers->at(0).first.code;
-        using P = multi::Protocol::Code;
-        if (protocol == P::WS) {
-          s << "/ws";
-        } else if (protocol == P::WSS) {
-          s << "/wss";
+      case P::DNS:
+      case P::DNS4:
+      case P::DNS6: {
+        Dns dns{std::nullopt, v};
+        switch (p.code) {
+          case P::DNS4:
```

Comment:
The case `P::DNS` is not covered by the inner switch.

Reply:
Probably it should be:
```suggestion
          case P::DNS:
          case P::DNS4:
```

Reply:
No, `v4 = match { DNS => None, DNS4 => Some(true), DNS6 => Some(false) }`